src/charm.py

```diff
@@ -75,26 +108,109 @@ def _on_postgresql_pebble_ready(self, event: WorkloadEvent) -> None:
+        try:
+            client = Client()
+            patch = {
+                "metadata": {"labels": {"application": "patroni", "cluster-name": self._namespace}}
+            }
+            client.patch(
+                Pod,
+                name=self._unit_name_to_pod_name(self._unit),
+                namespace=self._namespace,
+                obj=patch,
+            )
+        except ApiError as e:
+            logger.error("failed to patch pod")
+            self.unit.status = BlockedStatus(f"failed to patch pod with error {e}")
+
+    def _create_resources(self):
+        """Create kubernetes resources needed for Patroni."""
+        try:
+            client = Client()
+            with open("src/resources.yaml") as f:
+                for obj in codecs.load_all_yaml(f, context=self._context):
+                    client.create(obj)
+        except ApiError as e:
+            logger.error("failed to create resources")
+            self.unit.status = BlockedStatus(f"failed to create resources with error {e}")
+
     def _on_get_postgres_password(self, event: ActionEvent) -> None:
         """Returns the password for the postgres user as an action response."""
         event.set_results({"postgres-password": self._get_postgres_password()})
 
+    def _on_get_primary(self, event: ActionEvent) -> None:
+        """Get primary instance."""
+        try:
+            primary = self._patroni.get_primary(unit_name_pattern=True)
+            event.set_results({"primary": primary})
+        except RetryError as e:
+            logger.error(f"failed to get primary with error {e}")
+
+    def _on_update_status(self, _) -> None:
+        # Until https://github.com/canonical/pebble/issues/6 is fixed,
+        # we need to use the logic below to restart the leader
+        # and a stuck replica after a failover/switchover.
+        try:
+            state = self._patroni.get_postgresql_state()
+            if state == "restarting":
+                # Restart the stuck replica.
+                self._restart_postgresql_service(wait=False)
+            elif state == "stopping":
+                # Restart the stuck previous leader.
+                self._restart_postgresql_service()
+        except RetryError as e:
+            logger.error("failed to check PostgreSQL state")
+            self.unit.status = BlockedStatus(f"failed to check PostgreSQL state with error {e}")
+
+    def _restart_postgresql_service(self, wait=True) -> None:
+        """Restart PostgreSQL and Patroni.
+
+        Args:
+            wait: whether or not to wait 30 seconds between stop and start.
+        """
+        self.unit.status = MaintenanceStatus(f"restarting {self._postgresql_service} service")
+        container = self.unit.get_container("postgresql")
+        container.stop(self._postgresql_service)
+        # Sleep needed to make the leader release the lock in a failover/switchover.
+        if wait:
+            sleep(30)
```

Comment:
Hmm - I don't like this at all - sleeping in an hook blocks all other hook invocations/model operations for this charm. You could use `tenacity` here to just block on a specific condition if it's a must, but arbitrary sleeps for 30s is not a good look.

Reply:
Thanks for the feedback Jon! I definitely agree. I'll improve that with your suggestion to avoid blocking the other hooks.